lib/Audio/StreamGenerator.pm

```diff
@@ -248,34 +251,73 @@ sub _mix {
             if ( $value > $max[$channel] ) {
                 $max[$channel] = $value;
             }
+            if (!defined $first_loud_sample_index && $value > MAXINT) {
+                $first_loud_sample_index = $index;
+            }
         }
     }
 
     $self->debug( "done mixing" );
     croak "unused samples left in the buffer of the new track after mixing - this should never happen!" if @new_buffer;
 
-    # re-add the samples we may have skipped before mixing to the beginning of the buffer.
-    unshift( @$buffer, @skipped_buffer );
+    # The next step will be volume adjustment, so skip everything up to (excluding) the first loud sample.
+
+    my $to_skip = $first_loud_sample_index // @$buffer;
```

Comment:
Seems like `$first_loud_sample_index` should be incremented by 1 here? Currently, it takes one less element in splice below. (and `$first_loud_sample_index == 0` will splice nothing).

Probably better to use `$#$buffer` instead of `@$buffer` and then increment `$to_skip` (to keep current `$a // $b` construction intact)

Reply:
This is what I was thinking when I wrote this:

`$to_skip` is being passed as the `LENGTH` argument to splice - so it's supposed to contain the number of items that need to be removed. It's not an index. 

If the very first sample (index 0) in @$buffer is a loud sample, `$first_loud_sample_index` will be 0, which will cause no items to be spliced. Which is good - we want the first loud sample to stay in @$buffer. 

If the second sample (index 1) in @$buffer is the first loud sample, `$first_loud_sample_index` will be 1. Which will cause the first item (with index 0) to be spliced. Which to me again seems to be exactly what we want. 

If there are no loud samples, the whole buffer is 'not loud' and thus to be spliced and pushed to @skipped buffer, leaving an amount of 0 samples that need volume adjustment in @$buffer. 

Does this make sense? Please do correct me if my logic is flawed! :)
..

I was completely unaware of the existence of the `$#$buffer` functionality - thank you for pointing me to it! I guess i'm so used to doing things like `foreach (@bla)` that I never really got exposed to much code that does explicit stuff with array indexes. 

I think it still makes sense to keep useing @$buffer in this specific because as far as i understand we need the amount of samples there, not the index. But there are a few places where stuff like `for my $channel (0 .. @$sample - 1) {` is happening - these could probably benefit from using $#$sample instead.

Reply:
Okay, it seemed like you're mixing indexes with lengths here, but if this is correct then there's no problem ;)

I don't think an off by one sample error would make any difference in how mixing output sounds anyway, but just seemed a bit off.